### PR TITLE
Make Statement::sameAs virtual

### PR DESCRIPTION
**Pull request by naoyam**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 1]
```diff
@@ -176,16 +176,27 @@ Expr::Expr(const Expr* src, IrCloner* ir_cloner)
...
+      outputs().size() != other_expr->outputs().size()) {
     return false;
+  }
   for (size_t i = 0; i < inputs().size(); i++) {
-    if (!input(i)->sameAs(other->input(i)))
+    if (!input(i)->sameAs(other_expr->input(i))) {
       return false;
+    }
   }
+  // TODO: Why outputs are not checked?
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Why outputs are not checked?

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 2]
```diff
@@ -246,8 +294,17 @@ BroadcastOp::BroadcastOp(const BroadcastOp* src, IrCloner* ir_cloner)
...
-  return other->in() == in() && other->out() == out();
+bool BroadcastOp::sameAs(const Statement* other) const {
+  if (this == other) {
+    return true;
+  }
+  if (!other->isA<BroadcastOp>()) {
+    return false;
+  }
+  const auto other_op = other->as<BroadcastOp>();
+  // TODO: Why sameAs isn't used here?
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Why sameAs isn't used here?

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 3]
```diff
@@ -246,8 +294,17 @@ BroadcastOp::BroadcastOp(const BroadcastOp* src, IrCloner* ir_cloner)
...
+  // TODO: Other Exprs do not check outputs. Why it is checked here?
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Other Exprs do not check outputs. Why it is checked here?

**Review by tlemo — APPROVED** [review]

**Review by csarofeen — APPROVED** [review]
LGTM

**Pull request merged**